src/JsonFromString.ts

```diff
@@ -0,0 +1,43 @@
+/**
+ * @since 0.5.14
+ */
+import * as t from 'io-ts'
+
+/**
+ * Copied from `fp-ts/Either` module.
+ *
+ * @since 0.5.14
+ */
+export type Json = boolean | number | string | null | JsonArray | JsonRecord
+
+/**
+ * @since 0.5.14
+ */
+export interface JsonRecord {
+  readonly [key: string]: Json
+}
```

Comment:
I was about to open an issue in the fp-ts repo to discuss this separate issue but this seems like a good opportunity to bring it up so we can make the fix here as well.

When I tried to use this type alongside `t.partial` I ran into issues:

```ts
JSONFromString.pipe(
  /*
  Argument of type 'PartialC<{ foo: StringC; }>' is not assignable to parameter of type 'Type<{ foo?: string | undefined; }, Json, unknown>'.
    Types of property 'encode' are incompatible.
      Type 'Encode<{ foo?: string | undefined; }, { foo?: string | undefined; }>' is not assignable to type 'Encode<{ foo?: string | undefined; }, Json>'.
        Type '{ foo?: string | undefined; }' is not assignable to type 'Json'.
          Type '{ foo?: string | undefined; }' is not assignable to type 'JsonRecord'.
            Property 'foo' is incompatible with index signature.
              Type 'string | undefined' is not assignable to type 'Json'.
              Type 'undefined' is not assignable to type 'Json'.
  */
  t.partial({ foo: t.string }),
);
```

I believe this is because TypeScript does not distinguish between missing and `undefined`: https://github.com/microsoft/TypeScript/issues/13195#issuecomment-775930240

I mentioned some workarounds in that comment. What do you think about using one of those workarounds both here and in `Either.parseJSON`? So something like:

```suggestion
export type JsonRecord = { readonly [Key in string]?: Json };
```

Here is a reduced test case for the same issue when using `Either.Json`:

```ts
type User = { name?: string };

const user: User = { name: 'bob' };

// Type 'undefined' is not assignable to type 'Json'.
const userJson: E.Json = user;
```

---

Separately, can I ask why the parameter type of `Either.stringifyJSON` is `unknown` instead of `Json`? I presume it's also to workaround the same underlying TS issue?

Reply:
both workarounds would allow invalid values though

```ts
export type Json = boolean | number | string | null | JsonArray | JsonRecord
export type JsonRecord = { readonly [Key in string]?: Json }
export interface JsonArray extends ReadonlyArray<Json> {}

const x: JsonRecord = { a: undefined } // no compiler error
```

> can I ask why the parameter type of Either.stringifyJSON is unknown instead of Json?

`stringifyJSON` was added in v2.0.0 while the `Json` type was added in v2.6.7, changing `unknown` to `Json` would be a breaking change.
Nevertheless its current signature looks more general, isn't it?

Reply:
> both workarounds would allow invalid values though

You're right, I'm just not sure how else to workaround this error. 🤔 It seems there is no good workaround.

For now I am having to do this:

```ts
type JsonWithWorkaround = E.Json | { readonly [Key in string]?: JsonWithWorkaround };
export const JSONFromStringWithWorkaround = JSONFromString as t.Type<
  JsonWithWorkaround,
  string,
  string
>;
```

> Nevertheless its current signature looks more general, isn't it?

That's true, I just thought it would be better to be strict so you can't make common mistakes such as:

```ts
// You probably meant to remove the undefined values or
// convert them to `null` before calling `stringifyJSON`:
E.stringifyJSON({ foo: undefined }, () => '')
```

or

```ts
import * as E from 'fp-ts/Either';
import { pipe } from 'fp-ts/function';

pipe(
    E.stringifyJSON(undefined, () => ''),
    E.fold(
        () => {},
        (v) => {
            // Type is `string`
            // `undefined` at runtime ❗️
            console.log(v);
        },
    ),
);

```

Reply:
Just a note to say the issue I originally described in this thread no longer reproduces in TypeScript 4.2 due to this change: https://devblogs.microsoft.com/typescript/announcing-typescript-4-2-rc/#relaxed-rules-between-optional-properties-and-string-index-signatures (https://github.com/microsoft/TypeScript/pull/41921). 🎉